src/Symfony/Component/DependencyInjection/Definition.php

```diff
@@ -107,6 +107,10 @@ public function setFactory($factory)
             $factory = explode('::', $factory, 2);
         }
 
+        if ($factory instanceof Reference) {
```

Comment:
Could be `elseif`. If `\is_string($factory)` resolves to true, `instanceof` will never be true. Although it does not really matter, it seems to be more logical to me, WDYT?

Reply:
Yes it should be an elseif logically to avoid an extra if condition.
Also I would suggest to extract the duplicate parsing logic in setConfigurator and setFactory into a private method.